derp/derphttp/derphttp_client.go

```diff
@@ -583,6 +616,18 @@ func (c *Client) dialNode(ctx context.Context, n *tailcfg.DERPNode) (net.Conn, e
 	startDial := func(dstPrimary, proto string) {
 		nwait++
 		go func() {
+			if proto == "tcp4" && c.preferIPv6() {
```

Comment:
Why can't we use net.Dialer instead of all this machinery? Doesn't it already have logic to do happy eyeball racing?

I'm sure there's a good reason, I just can't see it with the tiny amount of context that github offers.

Reply:
Our DERP dialer has all those tailcfg/derpmap.go knobs,

https://pkg.go.dev/tailscale.com/tailcfg#DERPNode

... that let us/users control much more than what net.Dialer allows. It lets us do DNS or no DNS (hard-code v4 and/or v6 addresses), and control which cert is checked (which might different from the DNS lookup for domain fronting).

There's no way to use net.Dialer and retain all those knobs. But fortunately there's not much to net.Dialer's implementation that's not what this is doing, racing some goroutines.